The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16391337995/job/46317739263) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/46317739263) [(plain)](https://github.com/rust-lang/rust/commit/31b35e06833f1361e988137f81d0cb03e31ce86c/checks/46317739263/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/shims/x86/intrinsics-x86-avx2.rs ... ok
tests/pass/shims/x86/intrinsics-x86-gfni.rs ... ok

FAILED TEST: tests/pass/intrinsics/type-id.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-8n3Uki" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/pass/intrinsics" "tests/pass/intrinsics/type-id.rs" "--edition" "2021"

error: test got exit status: 101, but expected 0
 = note: the compiler panicked

error: no output was expected
---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::alloc_addresses::EvalContextExt>::addr_from_alloc_id
   4: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_alloc_root_pointer
   5: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::global_root_pointer
   6: <rustc_middle::mir::interpret::allocation::Allocation>::adjust_from_tcx::<miri::machine::Provenance, miri::alloc::alloc_bytes::MiriAllocBytes, rustc_middle::query::plumbing::TyCtxtAt, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#0}, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#1}>
   7: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation
   8: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_global_alloc
   9: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_alloc_raw
  10: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_ptr_alloc
  11: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_from_mplace_raw
  12: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_raw::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>>
  13: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::copy_op_no_validate::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>, rustc_const_eval::interpret::place::PlaceTy<miri::machine::Provenance>>
  14: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::eval_rvalue_into_place
  15: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::concurrency::thread::EvalContextExt>::run_threads
  16: miri::eval::eval_entry
  17: <miri::MiriCompilerCalls as rustc_driver_impl::Callbacks>::after_analysis
  18: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  19: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  20: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  22: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.
---
   = note: inside `std::any::TypeId::of::<usize>` at /checkout/library/core/src/any.rs:778:9: 778:45
note: inside `main`
  --> tests/pass/intrinsics/type-id.rs:7:14
   |
LL |     let t3 = TypeId::of::<usize>();
   |              ^^^^^^^^^^^^^^^^^^^^^
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /checkout/library/core/src/ops/function.rs:253:5: 253:71
   = note: inside `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /checkout/library/std/src/sys/backtrace.rs:158:18: 158:21
   = note: inside closure at /checkout/library/std/src/rt.rs:206:18: 206:75
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /checkout/library/core/src/ops/function.rs:290:13: 290:31
   = note: inside `std::panicking::catch_unwind::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside closure at /checkout/library/std/src/rt.rs:175:24: 175:49
   = note: inside `std::panicking::catch_unwind::do_call::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<isize, {closure@std::rt::lang_start_internal::{closure#0}}>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside `std::rt::lang_start_internal` at /checkout/library/std/src/rt.rs:171:5: 193:7
   = note: inside `std::rt::lang_start::<()>` at /checkout/library/std/src/rt.rs:205:5: 210:6



---
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::alloc_addresses::EvalContextExt>::addr_from_alloc_id
   4: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_alloc_root_pointer
   5: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::global_root_pointer
   6: <rustc_middle::mir::interpret::allocation::Allocation>::adjust_from_tcx::<miri::machine::Provenance, miri::alloc::alloc_bytes::MiriAllocBytes, rustc_middle::query::plumbing::TyCtxtAt, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#0}, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#1}>
   7: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation
   8: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_global_alloc
   9: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_alloc_raw
  10: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_ptr_alloc
  11: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_from_mplace_raw
  12: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_raw::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>>
  13: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::copy_op_no_validate::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>, rustc_const_eval::interpret::place::PlaceTy<miri::machine::Provenance>>
  14: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::eval_rvalue_into_place
  15: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::concurrency::thread::EvalContextExt>::run_threads
  16: miri::eval::eval_entry
  17: <miri::MiriCompilerCalls as rustc_driver_impl::Callbacks>::after_analysis
  18: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  19: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  20: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  22: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.
---
   = note: inside `std::any::TypeId::of::<usize>` at /checkout/library/core/src/any.rs:778:9: 778:45
note: inside `main`
  --> tests/pass/intrinsics/type-id.rs:7:14
   |
LL |     let t3 = TypeId::of::<usize>();
   |              ^^^^^^^^^^^^^^^^^^^^^
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at /checkout/library/core/src/ops/function.rs:253:5: 253:71
   = note: inside `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at /checkout/library/std/src/sys/backtrace.rs:158:18: 158:21
   = note: inside closure at /checkout/library/std/src/rt.rs:206:18: 206:75
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at /checkout/library/core/src/ops/function.rs:290:13: 290:31
   = note: inside `std::panicking::catch_unwind::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside closure at /checkout/library/std/src/rt.rs:175:24: 175:49
   = note: inside `std::panicking::catch_unwind::do_call::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<isize, {closure@std::rt::lang_start_internal::{closure#0}}>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside `std::rt::lang_start_internal` at /checkout/library/std/src/rt.rs:171:5: 193:7
   = note: inside `std::rt::lang_start::<()>` at /checkout/library/std/src/rt.rs:205:5: 210:6


full stdout:



FAILED TEST: tests/pass/panic/catch_panic.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-8n3Uki" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/pass/panic" "tests/pass/panic/catch_panic.rs" "--edition" "2021"

error: test got exit status: 101, but expected 0
 = note: the compiler panicked

error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/panic/catch_panic.stderr` to the actual output
--- tests/pass/panic/catch_panic.stderr
+++ <stderr output>
 
 thread 'main' panicked at tests/pass/panic/catch_panic.rs:LL:CC:
 Hello from std::panic
 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
 note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect
-Caught panic message (&str): Hello from std::panic
 
-thread 'main' panicked at tests/pass/panic/catch_panic.rs:LL:CC:
-Hello from std::panic: 1
-Caught panic message (String): Hello from std::panic: 1
+thread 'rustc' panicked at src/tools/miri/src/alloc_addresses/mod.rs:LL:CC:
+called `Result::unwrap_err()` on an `Ok` value: 0
+stack backtrace:
+   0: __rustc::rust_begin_unwind
+   1: core::panicking::panic_fmt
+   2: core::result::unwrap_failed
+   3: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::alloc_addresses::EvalContextExt>::addr_from_alloc_id
+   4: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_alloc_root_pointer
+   5: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::global_root_pointer
+   6: <rustc_middle::mir::interpret::allocation::Allocation>::adjust_from_tcx
+   7: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation
+   8: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_global_alloc
+   9: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_alloc_raw
+  10: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_ptr_alloc
+  11: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_from_mplace_raw
+  12: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_raw
+  13: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::copy_op_no_validate
+  14: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::eval_rvalue_into_place
+  15: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::concurrency::thread::EvalContextExt>::run_threads
+  16: miri::eval::eval_entry
+  17: <miri::MiriCompilerCalls as rustc_driver_impl::Callbacks>::after_analysis
+  18: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with
+  19: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt
+  20: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
+  21: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
+  22: rustc_interface::passes::create_and_enter_global_ctxt
+  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
 
-thread 'main' panicked at tests/pass/panic/catch_panic.rs:LL:CC:
---
+LL |         const { intrinsics::type_id::<T>() }
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: BACKTRACE:
+   = note: inside `std::any::TypeId::of::<std::string::String>` at RUSTLIB/core/src/any.rs:LL:CC
+   = note: inside `<(dyn std::any::Any + 'static)>::is::<std::string::String>` at RUSTLIB/core/src/any.rs:LL:CC
+   = note: inside `<(dyn std::any::Any + 'static)>::downcast_ref::<std::string::String>` at RUSTLIB/core/src/any.rs:LL:CC
+   = note: inside `<(dyn std::any::Any + std::marker::Send + 'static)>::downcast_ref::<std::string::String>` at RUSTLIB/core/src/any.rs:LL:CC
+note: inside `test::<{closure@tests/pass/panic/catch_panic.rs:LL:CC}>`
+  --> tests/pass/panic/catch_panic.rs:LL:CC
+   |
+LL |     let msg = if let Some(s) = res.downcast_ref::<String>() {
+   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+note: inside `main`
+  --> tests/pass/panic/catch_panic.rs:LL:CC
+   |
+LL |     test(None, |_old_val| std::panic!("Hello from std::panic"));
---
+   = note: inside `std::panicking::catch_unwind::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at RUSTLIB/std/src/panicking.rs:LL:CC
+   = note: inside `std::panicking::catch_unwind::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at RUSTLIB/std/src/panicking.rs:LL:CC
+   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at RUSTLIB/std/src/panic.rs:LL:CC
+   = note: inside closure at RUSTLIB/std/src/rt.rs:LL:CC
+   = note: inside `std::panicking::catch_unwind::do_call::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at RUSTLIB/std/src/panicking.rs:LL:CC
+   = note: inside `std::panicking::catch_unwind::<isize, {closure@std::rt::lang_start_internal::{closure#0}}>` at RUSTLIB/std/src/panicking.rs:LL:CC
+   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at RUSTLIB/std/src/panic.rs:LL:CC
+   = note: inside `std::rt::lang_start_internal` at RUSTLIB/std/src/rt.rs:LL:CC
+   = note: inside `std::rt::lang_start::<()>` at RUSTLIB/std/src/rt.rs:LL:CC
 
-thread 'main' panicked at tests/pass/panic/catch_panic.rs:LL:CC:
-assertion failed: false
---

thread 'main' panicked at tests/pass/panic/catch_panic.rs:50:27:
Hello from std::panic
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect

thread 'rustc' panicked at src/tools/miri/src/alloc_addresses/mod.rs:316:26:
called `Result::unwrap_err()` on an `Ok` value: 0
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::alloc_addresses::EvalContextExt>::addr_from_alloc_id
   4: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_alloc_root_pointer
   5: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::global_root_pointer
   6: <rustc_middle::mir::interpret::allocation::Allocation>::adjust_from_tcx::<miri::machine::Provenance, miri::alloc::alloc_bytes::MiriAllocBytes, rustc_middle::query::plumbing::TyCtxtAt, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#0}, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#1}>
   7: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation
   8: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_global_alloc
   9: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_alloc_raw
  10: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_ptr_alloc
  11: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_from_mplace_raw
  12: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_raw::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>>
  13: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::copy_op_no_validate::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>, rustc_const_eval::interpret::place::PlaceTy<miri::machine::Provenance>>
  14: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::eval_rvalue_into_place
  15: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::concurrency::thread::EvalContextExt>::run_threads
  16: miri::eval::eval_entry
  17: <miri::MiriCompilerCalls as rustc_driver_impl::Callbacks>::after_analysis
  18: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  19: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  20: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  22: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.
---
LL |         const { intrinsics::type_id::<T>() }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: BACKTRACE:
   = note: inside `std::any::TypeId::of::<std::string::String>` at /checkout/library/core/src/any.rs:778:9: 778:45
   = note: inside `<(dyn std::any::Any + 'static)>::is::<std::string::String>` at /checkout/library/core/src/any.rs:195:17: 195:34
   = note: inside `<(dyn std::any::Any + 'static)>::downcast_ref::<std::string::String>` at /checkout/library/core/src/any.rs:226:12: 226:26
   = note: inside `<(dyn std::any::Any + std::marker::Send + 'static)>::downcast_ref::<std::string::String>` at /checkout/library/core/src/any.rs:377:9: 377:43
note: inside `test::<{closure@tests/pass/panic/catch_panic.rs:50:16: 50:26}>`
  --> tests/pass/panic/catch_panic.rs:103:32
   |
LL |     let msg = if let Some(s) = res.downcast_ref::<String>() {
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main`
  --> tests/pass/panic/catch_panic.rs:50:5
   |
LL |     test(None, |_old_val| std::panic!("Hello from std::panic"));
---
   = note: inside `std::panicking::catch_unwind::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside closure at /checkout/library/std/src/rt.rs:175:24: 175:49
   = note: inside `std::panicking::catch_unwind::do_call::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<isize, {closure@std::rt::lang_start_internal::{closure#0}}>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside `std::rt::lang_start_internal` at /checkout/library/std/src/rt.rs:171:5: 193:7
   = note: inside `std::rt::lang_start::<()>` at /checkout/library/std/src/rt.rs:205:5: 210:6



full stderr:

thread 'main' panicked at tests/pass/panic/catch_panic.rs:50:27:
Hello from std::panic
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect

thread 'rustc' panicked at src/tools/miri/src/alloc_addresses/mod.rs:316:26:
called `Result::unwrap_err()` on an `Ok` value: 0
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::alloc_addresses::EvalContextExt>::addr_from_alloc_id
   4: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_alloc_root_pointer
   5: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::global_root_pointer
   6: <rustc_middle::mir::interpret::allocation::Allocation>::adjust_from_tcx::<miri::machine::Provenance, miri::alloc::alloc_bytes::MiriAllocBytes, rustc_middle::query::plumbing::TyCtxtAt, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#0}, <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation::{closure#1}>
   7: <miri::machine::MiriMachine as rustc_const_eval::interpret::machine::Machine>::adjust_global_allocation
   8: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_global_alloc
   9: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_alloc_raw
  10: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::get_ptr_alloc
  11: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_from_mplace_raw
  12: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::read_immediate_raw::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>>
  13: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::copy_op_no_validate::<rustc_const_eval::interpret::operand::OpTy<miri::machine::Provenance>, rustc_const_eval::interpret::place::PlaceTy<miri::machine::Provenance>>
  14: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine>>::eval_rvalue_into_place
  15: <rustc_const_eval::interpret::eval_context::InterpCx<miri::machine::MiriMachine> as miri::concurrency::thread::EvalContextExt>::run_threads
  16: miri::eval::eval_entry
  17: <miri::MiriCompilerCalls as rustc_driver_impl::Callbacks>::after_analysis
  18: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  19: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  20: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  21: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  22: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
  23: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.
---
LL |         const { intrinsics::type_id::<T>() }
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: BACKTRACE:
   = note: inside `std::any::TypeId::of::<std::string::String>` at /checkout/library/core/src/any.rs:778:9: 778:45
   = note: inside `<(dyn std::any::Any + 'static)>::is::<std::string::String>` at /checkout/library/core/src/any.rs:195:17: 195:34
   = note: inside `<(dyn std::any::Any + 'static)>::downcast_ref::<std::string::String>` at /checkout/library/core/src/any.rs:226:12: 226:26
   = note: inside `<(dyn std::any::Any + std::marker::Send + 'static)>::downcast_ref::<std::string::String>` at /checkout/library/core/src/any.rs:377:9: 377:43
note: inside `test::<{closure@tests/pass/panic/catch_panic.rs:50:16: 50:26}>`
  --> tests/pass/panic/catch_panic.rs:103:32
   |
LL |     let msg = if let Some(s) = res.downcast_ref::<String>() {
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main`
  --> tests/pass/panic/catch_panic.rs:50:5
   |
LL |     test(None, |_old_val| std::panic!("Hello from std::panic"));
---
   = note: inside `std::panicking::catch_unwind::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside closure at /checkout/library/std/src/rt.rs:175:24: 175:49
   = note: inside `std::panicking::catch_unwind::do_call::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panicking.rs:589:40: 589:43
   = note: inside `std::panicking::catch_unwind::<isize, {closure@std::rt::lang_start_internal::{closure#0}}>` at /checkout/library/std/src/panicking.rs:552:19: 552:88
   = note: inside `std::panic::catch_unwind::<{closure@std::rt::lang_start_internal::{closure#0}}, isize>` at /checkout/library/std/src/panic.rs:359:14: 359:40
   = note: inside `std::rt::lang_start_internal` at /checkout/library/std/src/rt.rs:171:5: 193:7
   = note: inside `std::rt::lang_start::<()>` at /checkout/library/std/src/rt.rs:205:5: 210:6


full stdout:
---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.29.2/src/lib.rs:369

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-0bacf351ef83efc6` (exit status: 1)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:38:22
  local time: Sat Jul 19 18:51:13 UTC 2025
  network time: Sat, 19 Jul 2025 18:51:13 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_const_eval/src/const_eval/machine.rs
+++ b/compiler/rustc_const_eval/src/const_eval/machine.rs
@@ -279,23 +279,15 @@ impl<'tcx> CompileTimeInterpCx<'tcx> {
     fn guaranteed_cmp(&mut self, a: Scalar, b: Scalar) -> InterpResult<'tcx, u8> {
         interp_ok(match (a, b) {
             // Comparisons between integers are always known.
-            (Scalar::Int { .. }, Scalar::Int { .. }) => {
-                if a == b {
-                    1
-                } else {
-                    0
-                }
-            }
-            // Comparisons of abstract pointers with null pointers are known if the pointer
-            // is in bounds, because if they are in bounds, the pointer can't be null.
-            // Inequality with integers other than null can never be known for sure.
-            (Scalar::Int(int), ptr @ Scalar::Ptr(..))
-            | (ptr @ Scalar::Ptr(..), Scalar::Int(int))
+            (Scalar::Int(a), Scalar::Int(b)) => (a == b) as u8,
+            // Comparisons of null with an arbitrary scalar can be known if `scalar_may_be_null`
+            // indicates that the scalar can definitely *not* be null.
+            (Scalar::Int(int), ptr) | (ptr, Scalar::Int(int))
                 if int.is_null() && !self.scalar_may_be_null(ptr)? =>
             {
                 0
             }
-            // Equality with integers can never be known for sure.
+            // Other ways of comparing integers and pointers can never be known for sure.
             (Scalar::Int { .. }, Scalar::Ptr(..)) | (Scalar::Ptr(..), Scalar::Int { .. }) => 2,
             // FIXME: return a `1` for when both sides are the same pointer, *except* that
             // some things (like functions and vtables) do not have stable addresses

--- a/compiler/rustc_const_eval/src/interpret/memory.rs
+++ b/compiler/rustc_const_eval/src/interpret/memory.rs
@@ -67,8 +67,10 @@ pub enum AllocKind {
     LiveData,
     /// A function allocation (that fn ptrs point to).
     Function,
-    /// A "virtual" allocation, used for vtables and TypeId.
-    Virtual,
+    /// A vtable allocation.
+    VTable,
+    /// A TypeId allocation.
+    TypeId,
     /// A dead allocation.
     Dead,
 }
@@ -952,7 +954,8 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             let kind = match global_alloc {
                 GlobalAlloc::Static { .. } | GlobalAlloc::Memory { .. } => AllocKind::LiveData,
                 GlobalAlloc::Function { .. } => bug!("We already checked function pointers above"),
-                GlobalAlloc::VTable { .. } | GlobalAlloc::TypeId { .. } => AllocKind::Virtual,
+                GlobalAlloc::VTable { .. } => AllocKind::VTable,
+                GlobalAlloc::TypeId { .. } => AllocKind::TypeId,
             };
             return AllocInfo::new(size, align, kind, mutbl);
         }
@@ -1617,6 +1620,13 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 match self.ptr_try_get_alloc_id(ptr, 0) {
                     Ok((alloc_id, offset, _)) => {
                         let info = self.get_alloc_info(alloc_id);
+                        if matches!(info.kind, AllocKind::TypeId) {
+                            // We *could* actually precisely answer this question since here,
+                            // the offset *is* the integer value. But the entire point of making
+                            // this a pointer is not to leak the integer value, so we say everything
+                            // might be null.
+                            return interp_ok(true);
+                        }
                         // If the pointer is in-bounds (including "at the end"), it is definitely not null.
                         if offset <= info.size {
                             return interp_ok(false);

--- a/src/tools/miri/src/alloc_addresses/mod.rs
+++ b/src/tools/miri/src/alloc_addresses/mod.rs
@@ -116,20 +116,25 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let this = self.eval_context_ref();
         let info = this.get_alloc_info(alloc_id);
 
-        // Miri's address assignment leaks state across thread boundaries, which is incompatible
-        // with GenMC execution. So we instead let GenMC assign addresses to allocations.
-        if let Some(genmc_ctx) = this.machine.data_race.as_genmc_ref() {
-            let addr = genmc_ctx.handle_alloc(&this.machine, info.size, info.align, memory_kind)?;
-            return interp_ok(addr);
-        }
-
-        let mut rng = this.machine.rng.borrow_mut();
         // This is either called immediately after allocation (and then cached), or when
         // adjusting `tcx` pointers (which never get freed). So assert that we are looking
         // at a live allocation. This also ensures that we never re-assign an address to an
         // allocation that previously had an address, but then was freed and the address
         // information was removed.
         assert!(!matches!(info.kind, AllocKind::Dead));
+
+        // TypeId allocations always have a "base address" of 0 (i.e., the relative offset is the
+        // hash fragment and therefore equal to the actual integer value).
+        if matches!(info.kind, AllocKind::TypeId) {
+            return interp_ok(0);
+        }
+
+        // Miri's address assignment leaks state across thread boundaries, which is incompatible
+        // with GenMC execution. So we instead let GenMC assign addresses to allocations.
+        if let Some(genmc_ctx) = this.machine.data_race.as_genmc_ref() {
+            let addr = genmc_ctx.handle_alloc(&this.machine, info.size, info.align, memory_kind)?;
+            return interp_ok(addr);
+        }
 
         // This allocation does not have a base address yet, pick or reuse one.
         if !this.machine.native_lib.is_empty() {
@@ -157,7 +162,7 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         this.get_alloc_bytes_unchecked_raw(alloc_id)?
                     }
                 }
-                AllocKind::Function | AllocKind::Virtual => {
+                AllocKind::Function | AllocKind::VTable => {
                     // Allocate some dummy memory to get a unique address for this function/vtable.
                     let alloc_bytes = MiriAllocBytes::from_bytes(
                         &[0u8; 1],
@@ -169,12 +174,13 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
                     std::mem::forget(alloc_bytes);
                     ptr
                 }
-                AllocKind::Dead => unreachable!(),
+                AllocKind::TypeId | AllocKind::Dead => unreachable!(),
             };
             // We don't have to expose this pointer yet, we do that in `prepare_for_native_call`.
             return interp_ok(base_ptr.addr().to_u64());
         }
         // We are not in native lib mode, so we control the addresses ourselves.
+        let mut rng = this.machine.rng.borrow_mut();
         if let Some((reuse_addr, clock)) = global_state.reuse.take_addr(
             &mut *rng,
             info.size,
@@ -295,21 +301,25 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 // Store address in cache.
                 global_state.base_addr.try_insert(alloc_id, base_addr).unwrap();
 
-                // Also maintain the opposite mapping in `int_to_ptr_map`, ensuring we keep it sorted.
-                // We have a fast-path for the common case that this address is bigger than all previous ones.
-                let pos = if global_state
-                    .int_to_ptr_map
-                    .last()
-                    .is_some_and(|(last_addr, _)| *last_addr < base_addr)
-                {
-                    global_state.int_to_ptr_map.len()
-                } else {
-                    global_state
+                // Also maintain the opposite mapping in `int_to_ptr_map`, ensuring we keep it
+                // sorted. We have a fast-path for the common case that this address is bigger than
+                // all previous ones. We skip this for allocations at address 0; those can't be
+                // real, they must be TypeId "fake allocations".
+                if base_addr != 0 {
+                    let pos = if global_state
                         .int_to_ptr_map
-                        .binary_search_by_key(&base_addr, |(addr, _)| *addr)
-                        .unwrap_err()
-                };
-                global_state.int_to_ptr_map.insert(pos, (base_addr, alloc_id));
+                        .last()
+                        .is_some_and(|(last_addr, _)| *last_addr < base_addr)
+                    {
+                        global_state.int_to_ptr_map.len()
+                    } else {
+                        global_state
+                            .int_to_ptr_map
+                            .binary_search_by_key(&base_addr, |(addr, _)| *addr)
+                            .unwrap_err()
+                    };
+                    global_state.int_to_ptr_map.insert(pos, (base_addr, alloc_id));
+                }
 
                 interp_ok(base_addr)
             }

--- a/src/tools/miri/src/borrow_tracker/stacked_borrows/mod.rs
+++ b/src/tools/miri/src/borrow_tracker/stacked_borrows/mod.rs
@@ -650,7 +650,7 @@ trait EvalContextPrivExt<'tcx, 'ecx>: crate::MiriInterpCxExt<'tcx> {
                         dcx.log_protector();
                     }
                 },
-                AllocKind::Function | AllocKind::Virtual | AllocKind::Dead => {
+                AllocKind::Function | AllocKind::VTable | AllocKind::TypeId | AllocKind::Dead => {
                     // No stacked borrows on these allocations.
                 }
             }
@@ -1021,7 +1021,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 trace!("Stacked Borrows tag {tag:?} exposed in {alloc_id:?}");
                 alloc_extra.borrow_tracker_sb().borrow_mut().exposed_tags.insert(tag);
             }
-            AllocKind::Function | AllocKind::Virtual | AllocKind::Dead => {
+            AllocKind::Function | AllocKind::VTable | AllocKind::TypeId | AllocKind::Dead => {
                 // No stacked borrows on these allocations.
             }
         }

--- a/src/tools/miri/src/borrow_tracker/tree_borrows/mod.rs
+++ b/src/tools/miri/src/borrow_tracker/tree_borrows/mod.rs
@@ -673,7 +673,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 trace!("Tree Borrows tag {tag:?} exposed in {alloc_id:?}");
                 alloc_extra.borrow_tracker_tb().borrow_mut().expose_tag(tag);
             }
-            AllocKind::Function | AllocKind::Virtual | AllocKind::Dead => {
+            AllocKind::Function | AllocKind::VTable | AllocKind::TypeId | AllocKind::Dead => {
                 // No tree borrows on these allocations.
             }
         }

